db/db.go

```diff
@@ -76,20 +78,64 @@ func Init(config config.DBConfig, debug bool) *Database {
 	if debug {
 		db.Config.Logger.LogMode(logger.Info)
 	}
-	return &Database{db}
+
+	d := &Database{db}
+	if err = d.MigrateLineage(); err != nil {
+		log.Fatalf("Lineage migration failed: %v\n", err)
+	}
+
+	return d
+}
+
+// MigrateLineage is a migration function to update db and its data to the
+// new lineage db scheme. It will update State table data, delete "lineage" column
+// and add corresponding Lineage entries
+func (db *Database) MigrateLineage() error {
+	if db.Migrator().HasColumn(&types.State{}, "lineage") {
+		states := db.ListStates()
+		for _, stPath := range states {
+			// Recover State from db for update
+			var st types.State
+			res := db.First(&st, types.State{Path: stPath})
+			if errors.Is(res.Error, gorm.ErrRecordNotFound) {
+				return fmt.Errorf("State not found in db")
+			}
+
+			if err := db.UpdateState(st); err != nil {
+				return fmt.Errorf("Failed to update %s state during lineage migration: %v", stPath, err)
+			}
+		}
+
+		// Custom migration rules
+		if err := db.Migrator().DropColumn(&types.State{}, "lineage"); err != nil {
+			return fmt.Errorf("Failed to drop lineage column during migration: %v", err)
+		}
+	}
+
+	return nil
 }
 
 type attributeValues map[string]interface{}
 
-func (db *Database) stateS3toDB(sf *statefile.File, path string, versionID string) (st types.State) {
+func (db *Database) stateS3toDB(sf *statefile.File, path string, versionID string) (st types.State, err error) {
 	var version types.Version
 	db.First(&version, types.Version{VersionID: versionID})
+
+	// Check if the associated lineage is already present in lineages table
+	// If so, it recovers its ID otherwise it inserts it at the same time as the state
+	var lineage types.Lineage
+	err = db.FirstOrCreate(&lineage, types.Lineage{Value: sf.Lineage}).Error
+	if err != nil || lineage.ID == 0 {
+		log.Error("Unknown error in stateS3toDB during lineage finding")
+		return types.State{}, err
+	}
+
 	st = types.State{
 		Path:      path,
 		Version:   version,
 		TFVersion: sf.TerraformVersion.String(),
 		Serial:    int64(sf.Serial),
-		Lineage:   sf.Lineage,
+		LineageID: lineage.ID,
```

Comment:
I wonder if we could use `Lineage: lineage` here, or even make gorm create the lineage automatically with an automatic junction

Reply:
I have finally removed this Lineage field, with the changes I made it became unnecessary.
But, indeed, I used it before to directly insert the Lineage at the same time as the State if it did not exist, the foreign key was also managed automatically. However, it was a problem when the Lineage already existed and I had to go through the foreign key directly ...
The current solution is simpler I find, I do a FirstOrCreate for the lineage first and after set the foreign key on state insertion.